### PR TITLE
fix FancyArrowPatch picker fails depending on arrowstyle

### PR DESCRIPTION
**Pull request by zhoubecky**
## PR Summary
fix FancyArrowPatch picker fails depending on arrowstyle #8384

### CODE REVIEW

**Comment by zhoubecky** [comment]
Every arrowstyle(Curve, Bracket or others) contains one to three sub paths which represent the head arrow (if present), line, and tail arrow (if present) respectively.  
If a Patch contains multiple Pathes, all the Pathes will be concatenated to one Path by joining the vertices list and the codes list. A composite Path with not properly closed sub Pathes will cause error in transformation thus resulting in incorrect evaluation of contains, a function check whether a point is within the Path.
FancyArrowPatch picker not works on Curve and Bracket arrows are because their paths consisted by multiple subpaths, and only the last subpath is closed properly(code 79). 

Because whether a location should respond to a mouse event or not depends on the result of the contains function. So I check whether any subpath contains the point separately(separate by the starting code=1) in contains function.

**Comment by zhoubecky** [comment]
Hi @tacaswell, I push a temp fix of this bug and add a test, could you take a look at it? Thank you.

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -137,9 +137,22 @@ def contains(self, mouseevent, radius=None):
...
-        inside = self.get_path().contains_point(
-            (mouseevent.x, mouseevent.y), self.get_transform(), radius)
+        codes = self.get_path().codes
+        vertices = self.get_path().vertices
+        for i in range(1, len(codes)+1):
+            # fix bug (#8384)
+            # if current path is concatenated by multiple pathes
+            # check if any subpath contains the point.
+            if (i == len(codes)) or (codes[i] == 1):
+                path = Path(np.concatenate([vertices[start: i]]),
```
Comment (tacaswell) [inline lib/matplotlib/patches.py]: Do you need `np.concatenate` here?

File: lib/matplotlib/patches.py [hunk 2]
```diff
@@ -137,9 +137,22 @@ def contains(self, mouseevent, radius=None):
...
+                            np.concatenate([codes[start: i]]))
+                inside = inside or path.contains_point(
```
Comment (tacaswell) [inline lib/matplotlib/patches.py]: once you have found that the point is in one path, I think we can short-circuit the loop and return?

File: lib/matplotlib/patches.py [hunk 3]
```diff
@@ -137,9 +137,22 @@ def contains(self, mouseevent, radius=None):
...
         if callable(self._contains):
             return self._contains(self, mouseevent)
+        inside = False
+        start = 0
         radius = self._process_radius(radius)
-        inside = self.get_path().contains_point(
-            (mouseevent.x, mouseevent.y), self.get_transform(), radius)
+        codes = self.get_path().codes
+        vertices = self.get_path().vertices
+        for i in range(1, len(codes)+1):
```
Comment (tacaswell) [inline lib/matplotlib/patches.py]: can you check that we do not already have something in the path module that breaks a path up into separate paths?

**Comment by tacaswell** [comment]
:+1: on sorting out what is wrong here!  This bug had me stumped.

File: lib/matplotlib/patches.py (hunk 3 above)
Comment (QuLogic) [reply lib/matplotlib/patches.py]: Yes, can we use `Path.iter_segments` here?

File: lib/matplotlib/patches.py (hunk 1 above)
Comment (zhoubecky) [reply lib/matplotlib/patches.py]: I will remove the np.concatenate.

File: lib/matplotlib/patches.py (hunk 2 above)
Comment (zhoubecky) [reply lib/matplotlib/patches.py]: ok will do.

File: lib/matplotlib/patches.py (hunk 3 above)
Comment (zhoubecky) [reply lib/matplotlib/patches.py]: Thanks you. I have tried the Path.iter_segments, it is not break the path to separate subpaths, but break it to every single vertice. I still need to do extra work to get the subpath(like create an empty vertices list to store each vertice or to find the index). Do you still want me to change to use Path.iter_segments?

**Comment by anntzer** [comment]
Something like
```
        radius = self._process_radius(radius)
        codes = self.get_path().codes
        vertices = self.get_path().vertices
        idxs, = np.nonzero(codes == 1)
        idxs = idxs[1:]  # Don't split before the first MOVETO.
        return any(
            subpath.contains_point(
                (mouseevent.x, mouseevent.y), self.get_transform(), radius)
            for subpath in map(
                Path, np.split(vertices, idxs), np.split(codes, idxs))), {}
```
is probably faster (moving all the logic to numpy and taking advantage of the lazy map).

But +1 to the PR in general.

**Comment by zhoubecky** [comment]
Hi @anntzer , thanks for your suggestion! I have updated the code.

File: lib/matplotlib/patches.py [hunk 4]
```diff
@@ -138,9 +138,16 @@ def contains(self, mouseevent, radius=None):
         if callable(self._contains):
             return self._contains(self, mouseevent)
         radius = self._process_radius(radius)
-        inside = self.get_path().contains_point(
-            (mouseevent.x, mouseevent.y), self.get_transform(), radius)
-        return inside, {}
+        codes = self.get_path().codes
+        vertices = self.get_path().vertices
+        idxs, = np.nonzero(codes == 1)
```
Comment (dstansby) [inline lib/matplotlib/patches.py]: Thanks for looking into this! What does this line do? Might be worth adding a comment just before it explaining.

Comment (QuLogic) [reply lib/matplotlib/patches.py]: `np.where` might be clearer.

**Comment by zhoubecky** [comment]
hi @tacaswell , is there anything else need to be fixed to make it better?

File: lib/matplotlib/tests/test_path.py [hunk 5]
```diff
@@ -65,6 +67,46 @@ def test_nonlinear_containment():
...
+    X, Y = np.meshgrid(np.arange(0, 2, 0.1),
+                       np.arange(0, 2, 0.1))
+    for k, (x, y) in enumerate(zip(X.ravel(), Y.ravel())):
+        xdisp, ydisp = ax.transData.transform_point([x, y])
+        event = MouseEvent('button_press_event', fig.canvas, xdisp, ydisp)
+        for m, patch in enumerate(patches_list):
+            # set the points to red only if the arrow contains the point
+            inside, res = patch.contains(event)
+            if inside:
+                plt.scatter(x, y, s=5, c="r")
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_path.py]: Should use `ax.scatter`

File: lib/matplotlib/patches.py [hunk 6]
```diff
@@ -138,9 +138,18 @@ def contains(self, mouseevent, radius=None):
...
+        # if the current path is concatenated by multiple sub paths.
+        # get the indexes of the starting code(MOVETO) of all sub paths
+        idxs, = np.where(codes == 1)
+        # Don't split before the first MOVETO.
+        idxs = idxs[1:]
+        return any(
+            subpath.contains_point(
+                (mouseevent.x, mouseevent.y), self.get_transform(), radius)
+            for subpath in map(
+                Path, np.split(vertices, idxs), np.split(codes, idxs))), {}
```
Comment (jklymak) [inline lib/matplotlib/patches.py]: I think I finally see what this is doing, but,  can this be deconvolved to ease in maintenance and understanding?  Other devs are free to approve if they are OK with this, and its just my python ignorance, but its not the clearest return statement in the code base.

**Comment by tacaswell** [comment]
Power-cycled to re-trigger CI

**Comment by tacaswell** [comment]
@zhoubecky Are you still interested in working on this and can you look into why the tests are failing?  I suspect a re-base may help.

**Comment by zhoubecky** [comment]
@tacaswell sorry I just saw your message. I will take a look this weekend.

**Comment by zhoubecky** [comment]
@tacaswell I have checked the failure message and it shows ImageComparisonFailure, the expected image name is arrow_contains_point-expected.png but the actual is arrow_contains_point.png. Change the image name and the tests should pass. I will double check and fix it. I need some time to reinstall the developing environment. Sorry for the late.

**Comment by tacaswell** [comment]
Power-cycled to re-run against current master.

File: lib/matplotlib/patches.py (hunk 6 above)
Comment (timhoffm) [reply lib/matplotlib/patches.py]: Could you reformat this to

~~~
        subpaths = map(Path, np.split(vertices, idxs), np.split(codes, idxs)))
        contains = any(
            subpath.contains_point(
                (mouseevent.x, mouseevent.y), self.get_transform(), radius)
            for subpath in subpaths))
	return contains, {}
~~~

File: lib/matplotlib/patches.py [hunk 7]
```diff
@@ -130,9 +130,18 @@ def contains(self, mouseevent, radius=None):
...
             return self._contains(self, mouseevent)
         radius = self._process_radius(radius)
-        inside = self.get_path().contains_point(
-            (mouseevent.x, mouseevent.y), self.get_transform(), radius)
-        return inside, {}
+        codes = self.get_path().codes
+        vertices = self.get_path().vertices
+        # if the current path is concatenated by multiple sub paths.
+        # get the indexes of the starting code(MOVETO) of all sub paths
+        idxs, = np.where(codes == 1)
```
Comment (timhoffm) [inline lib/matplotlib/patches.py]:
```suggestion
        idxs, = np.where(codes == Path.MOVETO)
```

File: lib/matplotlib/patches.py (hunk 6 above)
Comment (tacaswell) [reply lib/matplotlib/patches.py]: I think it is OK to leave as-is.

**Comment by tacaswell** [comment]
Thanks @zhoubecky !  Sorry this took almost a year to get in, hopefully we will hear from you again.

**Pull request merged**